### PR TITLE
Ensure finalizer is removed when 409 Conflict returned on status update

### LINKED ISSUE #2066: Finalizer not cleared (no backoff endless loop)
I just got the controller into a state where it continually tries to clear the finalizer:
```
I0524 10:53:25.697656       1 controller_instance.go:570] ServiceInstance "foo11/dh-postgresql-apb-wzxth": Processing deleting event
I0524 10:53:25.697684       1 controller_instance.go:1211] ServiceInstance "foo11/dh-postgresql-apb-wzxth": Updating status
I0524 10:53:25.888709       1 request.go:481] Throttling request took 190.866004ms, request: PUT:https://172.30.0.1:443/apis/servicecatalog.k8s.io/v1beta1/namespaces/foo11/serviceinstances/dh-postgresql-apb-wzxth/status
I0524 10:53:26.088753       1 request.go:481] Throttling request took 195.642467ms, request: GET:https://172.30.0.1:443/apis/servicecatalog.k8s.io/v1beta1/namespaces/foo11/serviceinstances/dh-postgresql-apb-wzxth
I0524 10:53:26.288708       1 request.go:481] Throttling request took 94.066335ms, request: PUT:https://172.30.0.1:443/apis/servicecatalog.k8s.io/v1beta1/namespaces/foo11/serviceinstances/dh-postgresql-apb-wzxth/status
I0524 10:53:26.295891       1 controller_instance.go:1639] ServiceInstance "foo11/dh-postgresql-apb-wzxth": Cleared finalizer
I0524 10:53:26.295934       1 controller_instance.go:570] ServiceInstance "foo11/dh-postgresql-apb-wzxth": Processing deleting event
I0524 10:53:26.295966       1 controller_instance.go:1211] ServiceInstance "foo11/dh-postgresql-apb-wzxth": Updating status
I0524 10:53:26.488723       1 request.go:481] Throttling request took 192.639103ms, request: PUT:https://172.30.0.1:443/apis/servicecatalog.k8s.io/v1beta1/namespaces/foo11/serviceinstances/dh-postgresql-apb-wzxth/status
I0524 10:53:26.688749       1 request.go:481] Throttling request took 195.108459ms, request: GET:https://172.30.0.1:443/apis/servicecatalog.k8s.io/v1beta1/namespaces/foo11/serviceinstances/dh-postgresql-apb-wzxth
I0524 10:53:26.888682       1 request.go:481] Throttling request took 94.082646ms, request: PUT:https://172.30.0.1:443/apis/servicecatalog.k8s.io/v1beta1/namespaces/foo11/serviceinstances/dh-postgresql-apb-wzxth/status
I0524 10:53:26.894821       1 controller_instance.go:1639] ServiceInstance "foo11/dh-postgresql-apb-wzxth": Cleared finalizer
I0524 10:53:26.894878       1 controller_instance.go:570] ServiceInstance "foo11/dh-postgresql-apb-wzxth": Processing deleting event
I0524 10:53:26.894914       1 controller_instance.go:1211] ServiceInstance "foo11/dh-postgresql-apb-wzxth": Updating status
I0524 10:53:27.088673       1 request.go:481] Throttling request took 193.586544ms, request: PUT:https://172.30.0.1:443/apis/servicecatalog.k8s.io/v1beta1/namespaces/foo11/serviceinstances/dh-postgresql-apb-wzxth/status
I0524 10:53:27.288781       1 request.go:481] Throttling request took 197.250665ms, request: GET:https://172.30.0.1:443/apis/servicecatalog.k8s.io/v1beta1/namespaces/foo11/serviceinstances/dh-postgresql-apb-wzxth
I0524 10:53:27.467299       1 leaderelection.go:199] successfully renewed lease kube-service-catalog/service-catalog-controller-manager
I0524 10:53:27.488659       1 request.go:481] Throttling request took 93.569546ms, request: PUT:https://172.30.0.1:443/apis/servicecatalog.k8s.io/v1beta1/namespaces/foo11/serviceinstances/dh-postgresql-apb-wzxth/status
I0524 10:53:27.492282       1 controller_instance.go:1639] ServiceInstance "foo11/dh-postgresql-apb-wzxth": Cleared finalizer
I0524 10:53:27.492311       1 controller_instance.go:570] ServiceInstance "foo11/dh-postgresql-apb-wzxth": Processing deleting event
```

I'll try to find out what's going on. Submitting issue just so I don't forget.

### PR DESCRIPTION
**Pull request by luksa**
Fixes #2066

### CODE REVIEW

File: pkg/controller/controller_instance.go [hunk 1]
```diff
@@ -1207,6 +1207,13 @@ func (c *controller) updateServiceInstanceReferences(toUpdate *v1beta1.ServiceIn
 // Note: objects coming from informers should never be mutated; the instance
 // passed to this method should always be a deep copy.
 func (c *controller) updateServiceInstanceStatus(instance *v1beta1.ServiceInstance) (*v1beta1.ServiceInstance, error) {
+	return c.updateServiceInstanceStatusWithRetries(instance, nil)
+}
+
+func (c *controller) updateServiceInstanceStatusWithRetries(
+	instance *v1beta1.ServiceInstance,
+	updateFunc func(*v1beta1.ServiceInstance)) (*v1beta1.ServiceInstance, error) {
```
Comment (jboyd01) [inline pkg/controller/controller_instance.go]: I'd certainly benefit from a comment on this function indicating the updateFunc (if specified) applies additional updates to the instance which must be applied in the event this instance is out of date and an updated instance must be used. 

LGTM though.

Comment (carolynvs) [reply pkg/controller/controller_instance.go]: Agreed, this really needs more explanation and perhaps a different name? If I understand this properly the function is applied only when retrying after an conflict, a more clear name would perhaps be `resolveConflictFunc` or something along those lines.

Comment (luksa) [reply pkg/controller/controller_instance.go]: I've added a comment & renamed the function to `postConflictUpdateFunc`.

**Pull request merged**